src/backend/api/meals.js

```diff
@@ -5,11 +5,167 @@ const knex = require("../database");
 router.get("/", async (request, response) => {
   try {
     // knex syntax for selecting things. Look up the documentation for knex for further info
-    const titles = await knex("meals").select("title");
-    response.json(titles);
+    
+    const titles = await knex("meal").select("title");
+    console.log(titles)
+    response.send(titles);
+
+  } catch (error) {
+    console.log(error.message);
+    response.status(500).send({ error: "Internal Server Error." });
+  }
+});
+
+router.get("/max-price", async (request, response) => {
```

Comment:
The idea is that the `GET /` route should support all the query parameters

- maxPrice
- availableReservations
- title
- createdAfter
- limit

so you shouldn't have the route `/max-price`, `/availableReservations`, etc.

Reply:
Then it would become a long list of if-then-elses in the router that in my opinion wouldn't be modular. I read somewhere that it's better to have separate end-points for various business logics. Here as I understand, we don't want to combine the parameters in our query if more than one is provided in the URL(it would be a much more complicated exercise then and nobody in our class has implemented it that way). Will it fall short in delivering the purpose the way I have implemented? Otherwise, it's a simple matter of putting all the logic in just one router each in an if block.

Reply:
> I read somewhere that it's better to have separate end-points for various business logics.

It depends on the use case. If you look at the experience on [skoringen.dk](https://www.skoringen.dk/search?searchTerm=ecco) you can see that we can filter by gender, shoe size, color, etc. This is similar to what you're doing here. It wouldn't be a good idea in that case to have separate endpoints for each type of filter. Instead you would typically have a single endpoint and send the chosen filters as query params.

> Otherwise, it's a simple matter of putting all the logic in just one router each in an if block.

It's not as simple as "putting all the logic" because the SQL query you generate should be able to support all combinations of the various query params.

Reply:
Thanks for quick feedback. I still need some clarification here.

 > It depends on the use case. If you look at the experience on [skoringen.dk](https://www.skoringen.dk/search?searchTerm=ecco) you can see that we can filter by gender, shoe size, color, etc. This is similar to what you're doing here. It wouldn't be a good idea in that case to have separate endpoints for each type of filter. Instead you would typically have a single endpoint and send the chosen filters as query params.

In the example website you provided above, it's possible to choose all combinations together (size, gender and color). Here as I mentioned, they seem to be separate and combinations of params are not needed. If needed, nobody's code that I know of in our class is supporting it, because everyone has implemented each query individually in a separate 'if' block. So params cannot affect each other. Pleaes correct me if I'm wrong. I really need to know if the single Get end-point should support the combination of request params or not (e.g. /meals?max-price=120&availableReservations=true). If so, I'm not sure how we can build up knex conditions dynamically programmatically. Do you have any examples? 

> It's not as simple as "putting all the logic" because the SQL query you generate should be able to support all combinations of the various query params.

Does anyone's solution in our class support combinations of various query params? I think I need some help here.

Reply:
You're right that it seems like most people from the class implemented support for only one query parameter at a time, not all combinations. The homework text could be clarified a bit more to mention that ideally you should be supporting all combinations of query params. Your solution is a bit different though because you have created separate endpoints which is not as described in the homework text.

So you can either
- do as the others from the class and have 1 endpoint that only supports 1 query param at a time,
- or do the more complex version where you have 1 endpoint that supports all combinations of query params.

I think I will create a video and share with you and the rest of the class how you can do the more complex version, as it might not be obvious.

Reply:
I managed to implement combining the query parameters by using knex.raw. I hope it meets your expectations.

Reply:
Nice work. I haven't tested it out but it looks more like what I was thinking. Make sure to test with different combinations of query params and see if you get the expected output.

Reply:
Thank you so much, Orhan, for your time and all of your help and support. I learned a lot with your help.